File: solutions/php/yacht/1/Yacht.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

class Yacht
{
    public function score(array $rolls, string $category): int
    {
        sort($rolls);
        $counts = array_count_values($rolls);
        arsort($counts); // Sort by frequency descending

        switch ($category) {
            case 'ones':
                return ($counts[1] ?? 0) * 1;
            
            case 'twos':
                return ($counts[2] ?? 0) * 2;
            
            case 'threes':
                return ($counts[3] ?? 0) * 3;
            
            case 'fours':
                return ($counts[4] ?? 0) * 4;
            
            case 'fives':
                return ($counts[5] ?? 0) * 5;
            
            case 'sixes':
                return ($counts[6] ?? 0) * 6;
            
            case 'full house':
                // Must have exactly two different numbers, with frequencies 3 and 2
                if (count($counts) === 2 && in_array(3, $counts) && in_array(2, $counts)) {
                    return array_sum($rolls);
                }
                return 0;
            
            case 'four of a kind':
                // At least four dice showing the same face. Score sum of those four.
                foreach ($counts as $face => $count) {
                    if ($count >= 4) {
                        return $face * 4;
                    }
                }
                return 0;
            
            case 'little straight':
                // Must be 1-2-3-4-5
                if ($rolls === [1, 2, 3, 4, 5]) {
                    return 30;
                }
                return 0;
            
            case 'big straight':
                // Must be 2-3-4-5-6
                if ($rolls === [2, 3, 4, 5, 6]) {
                    return 30;
                }
                return 0;
            
            case 'choice':
                return array_sum($rolls);
            
            case 'yacht':
                // All five dice showing the same face
                if (count($counts) === 1) {
                    return 50;
                }
                return 0;
            
            default:
                return 0;
        }
    }
}

File: solutions/ruby/yacht/1/yacht.rb
class Yacht
  def initialize(dice, category)
    @dice = dice
    @category = category
  end

  def score
    counts = @dice.tally

    case @category
    when 'ones'   then count_single(1)
    when 'twos'   then count_single(2)
    when 'threes' then count_single(3)
    when 'fours'  then count_single(4)
    when 'fives'  then count_single(5)
    when 'sixes'  then count_single(6)
    
    when 'full house'
      # Un Full House necesita que los dados agrupados tengan exactamente las frecuencias 3 y 2
      counts.values.sort == [2, 3] ? @dice.sum : 0

    when 'four of a kind'
      # Buscamos si alguna cara se repite 4 o 5 veces
      four_of_a_kind_face = counts.find { |_, v| v >= 4 }&.first
      four_of_a_kind_face ? four_of_a_kind_face * 4 : 0

    when 'little straight'
      @dice.sort == [1, 2, 3, 4, 5] ? 30 : 0

    when 'big straight'
      @dice.sort == [2, 3, 4, 5, 6] ? 30 : 0

    when 'choice'
      @dice.sum

    when 'yacht'
      # Yacht requiere que las 5 frecuencias apunten a una sola cara (frecuencia [5])
      counts.values == [5] ? 50 : 0

    else
      0
    end
  end

  private

  # Método auxiliar para las categorías numéricas simples (ones, twos, etc.)
  def count_single(target)
    @dice.count(target) * target
  end
end